nasa/ow_simulator
# Oceanwater 747 Fill Scoop with Sample when it Digs

## Linked Issues:
| EPIC ⚡ | [OCEANWATER-680](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-680) |
| :----------- | :----------- |
| Jira Ticket 🎟️ | [OCEANWATER-747](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-747) |
| Github :octocat:  | n/a |

https://user-images.githubusercontent.com/17039973/137366612-4da6d98b-72dd-448c-842a-c83bcc874a3c.mp4

## Summary of Changes

* Both the visual and collision Gazebo plug-ins defined in ow_dynamic_terrain now publish a differential image describing how terrain has been modified on /ow_dynamic_terrain/modification_differential/visual and /ow_dynamic_terrain/modification_differential/collision, respectively.
* API method calls that reinitialize terrain objects and physics and are required following a visual or collision terrain modification will now only ever occur if there was an actual change in height computed by the `merge_method`. Similarly, the modification_differential topic will only be published if a change of height has occurred. This has the benefit of significantly speeding up the simulation when arm digging and grinding actions are operating.
* A new package, ow_regolith, has been added which defines `regolith_node`, a ROS node, that tracks changes to the visual terrain via the modification_differential topic. When a threshold volume of terrain has been removed, a regolith model will be spawned into the scoop, and held there by a fake force until the dig operation completes. The package also handles clean-up of these regolith models by simply deleting all of them at the end of an deliver arm action. 

## Caveats 

* The node will not work properly if the user calls the ROS Service versions of arm operations, instead of their action counterparts. Keep this in mind when testing! Do not test with ROS services.
* While sample is in the scoop you may see the following errors flood the console `ODE Message 3: LCP internal error, s <= 0 (s=0.0000e+00)`. I will attempt to address this in [OCEANWATER-819](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-819)

## Test 1 - using default digging parameters
1. Launch the simulation to the `atacama_y1a` world settings the sim_regolith flag to true
```roslaunch ow atacama_y1a.launch sim_regolith:=true```
2. In another terminal call the grind action
```./src/ow_simulator/ow_lander/scripts/grind_action_client.py```
No material will spawn as a result of the grind operation, but ensure the operation performs nominally.
3. Call either the dig linear or dig circular actions
```./src/ow_simulator/ow_lander/scripts/dig_linear_action_client.py```
While digging occurs terrain will be modified, and every so often a black sphere should spawn in the scoop and get pushed to the back. Some may fall out, but most will be held in the scoop. When the dig action completes, the fake force holding the spheres should be release and you should see them settle to the bottom.
4. Call the deliver arm action.
```./src/ow_simulator/ow_lander/scripts/deliver_action_client.py```
The spheres should remain in the scoop up until the actual delivery. Do not worry if the scoops fail to fall into the sample dock, that's a problem outside of the scope of this PR. When the deliver action is complete, you should see the spheres be deleted. You can confirm this by looking in the models dropdown box in Gazebo and confirming all models with the name `regolith_#` are gone.

## Test 2 - using non-default digging parameters
1. Use non-default values by modifying `grind_action_client.py` and either `dig_circular_action_client.py` or `dig_linear_action_client.py` files located in `ow_lander/scripts`.  To change action parameters modify where they are defined under the `# Default trenching values` line. Careful not to pick values that will place the arm outside of the workspace. What follows are examples of how I modified my `grind_action_client.py` and `dig_circular_action_client.py` to perform my test.
```Python
    # in grind_action_client.py
    goal.x_start = 1.65
    goal.y_start = 0.5
    goal.depth = 0.05
    goal.length = 0.6 
    goal.parallel = False
    goal.ground_position = constants.DEFAULT_GROUND_HEIGHT
```
```Python
    # in dig_circular_action_client.py
    goal.x_start = 1.65
    goal.y_start = 0.5
    goal.depth = 0.01
    goal.parallel = False
    goal.ground_position = constants.DEFAULT_GROUND_HEIGHT
```
2. Repeat Test 1 using your modified scripts and calling either dig linear or dig circular (depending which you modified).
3. When finished, be sure to revert the action script changes you made with the following command
```Python 
git checkout ow_lander/scripts/dig_linear_action_client.py ow_lander/scripts/dig_circular_action_client.py ow_lander/scripts/grind_action_client.py
```
